app/use_cases/update_shopping_cart.rb

```diff
@@ -0,0 +1,36 @@
+class UpdateShoppingCart
+  def run!(current_items:, target_items:)
+    item_types_to_keep = []
+    target_items.each do |i|
+      item_types_to_keep << i.item_type
+    end
+
+    current_items.each do |item|
+      unless item_types_to_keep.include? item.item_type
+        item.destroy!
+      end
+    end
+
+    target_items.each do |item_to_add|
+      current_item = current_items.find { |i| i.item_type == item_to_add.item_type }
+      if current_item
+        if current_item.class == DiscountItem
```

Comment:
take a look on https://robots.thoughtbot.com/tell-dont-ask principle if it suits in this case